Objective-C/Internal/Replicator/CBLWebSocket.mm

```diff
@@ -323,6 +347,99 @@ - (void) _connect {
     }
 }
 
+- (BOOL) connectToHostWithName: (NSString*)hostname
+                          port: (NSInteger)port
+              networkInterface: (NSString* _Nullable)interface
+                   inputStream: (NSInputStream**)inputStream
+                  outputStream: (NSOutputStream**)outputStream
+                         error: (NSError** _Nullable)error
+{
+    struct addrinfo hints;
+    memset(&hints, 0, sizeof(hints));
+    hints.ai_socktype = SOCK_STREAM;
+    
+    struct addrinfo* addr;
+    const char* cHost = [hostname cStringUsingEncoding: NSUTF8StringEncoding];
+    const char* cPort = [[NSString stringWithFormat: @"%ld", (long)port]
+                         cStringUsingEncoding: NSUTF8StringEncoding];
+    if (getaddrinfo(cHost, cPort, &hints, &addr)) {
+        NSString* m = [NSString stringWithFormat: @"failed to get address info, %s", strerror(errno)];
+        return [self populateError: error errNo: errno msg: m];
+    }
+    
+    int sockfd = socket(addr->ai_family, addr->ai_socktype, addr->ai_protocol);
+    if (sockfd < 0) {
+        freeaddrinfo(addr);
+        
+        NSString* m = [NSString stringWithFormat: @"failed to create socket, %s", strerror(errno)];
+        return [self populateError: error errNo: errno msg: m];
+    }
+    
+    if (interface) {
+        unsigned int index = if_nametoindex([interface cStringUsingEncoding: NSUTF8StringEncoding]);
+        if (index == 0) {
+            freeaddrinfo(addr);
+            
+            NSString* m = [NSString stringWithFormat: @"failed to find network interface %@ %s",
+                           interface, strerror(errno)];
+            return [self populateError: error errNo: errno msg: m];
+        }
+        int result = -1;
+        switch (addr->ai_family) {
+            case AF_INET:
+                result = setsockopt(sockfd, IPPROTO_IP, IP_BOUND_IF, &index, sizeof(index));
+                break;
+            case AF_INET6:
+                result = setsockopt(sockfd, IPPROTO_IPV6, IPV6_V6ONLY, &index, sizeof(index));
+                break;
+            default:
+                CBLWarnError(WebSocket, @"%@: Address family not supported! %d",
+                             self, addr->ai_family);
+                result = -1;
+                break;
+        }
+        if (result < 0) {
+            freeaddrinfo(addr);
+            
+            NSString* m = [NSString stringWithFormat: @"failed to set network interface %@ %s",
```

Comment:
I think we should include the info about addr->ai_family, addr->ai_socktype, addr->ai_protocol here too.